File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <iostream>
using namespace std;

int N, S;
int v[21];
int result;

void dfs(int i, int sum) {
    if(i == N) return;

    if(sum + v[i] == S) result++;

    dfs(i + 1, sum);
    dfs(i + 1, sum + v[i]);
}

int main(void) {
    cin >> N >> S;

    for(int i = 0; i < N; i++) {
        cin >> v[i];
    }

    dfs(0, 0);

    cout << result << endl;

    return 0;
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] v;
    static int N, S, result = 0;

    static void dfs(int i, int sum) {
        if(i == N) {
            if(sum == S) result++;
            return;
        }

        dfs(i + 1, sum);
        dfs(i + 1, sum + v[i]);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        S = Integer.parseInt(st.nextToken());
        v = new int[N];

        st = new StringTokenizer(br.readLine(), " ");
        for(int i = 0; i < N; i++) {
            v[i] = Integer.parseInt(st.nextToken());
        }

        dfs(0, 0);

        if(S == 0) result--;
        System.out.println(result);

        bw.flush();
        bw.close();
        br.close();
    }
}